### ISSUE TITLE
Problems with null values in […] queries on a class

### ISSUE THREAD

**Issue by tobiemh** [orientechnologies/orientdb#3368]
There seems to be a problem with searching for null values in […] queries.

When I perform the following query, searching for a specific RID, the query works successfully, and all tags which do not have a ‘deleted_at’ property are successfully retrieved, along with the #21:0 record.

```sql
SELECT *, out('Has')[@class=Tag][deleted_at=null] AS tags FROM #21:0
```

When I perform the same query, but on the whole class, a java.lang.NullPointerException is returned, and neither the People or the tags are returned.

```sql
SELECT *, out('Has')[@class=Tag][deleted_at=null] AS tags FROM People
```

When I perform the query on the whole class, without the [deleted_at=null] part - it is successful (but obviously also returns tags which have a ‘deleted_at’ property set).

```sql
SELECT *, out('Has')[@class=Tag] AS tags FROM People
```

OrientDB version: 2.0-rc1

**Comment by tobiemh** [comment]
I've looked slightly deeper at this issue, and the problem is not to do with whether it is searching a single record, or a whole Class. Instead it causes a _NullPointerException_ if the record (or any record on the class) does not have any tags. The problem is that the pipeline filter searching for null values needs a set/array.

If you do the following query on a Person record who does not have any Tags connected to it, then the query will fail:
```SQL
SELECT *, out('Has')[@class=Tag][deleted_at=null] AS tags FROM #21:0
```

If you make sure that the [@class=Tag] returns a set, then the query is successful:

```SQL
SELECT *, out('Has')[@class=Tag].asSet()[deleted_at=null] AS tags FROM #21:0
```
or
```SQL
SELECT *, out('Has')[@class=Tag].asSet()[deleted_at=null] AS tags FROM People
```

OrientDB version: 2.0.2

**Comment by lvca** [comment]
Could you please attach the full stack trace?

**Comment by tobiemh** [comment]
The full stack trace is as follows:
```
Sent run-time exception to the client /172.28.128.1:62363: com.orientechnologies.orient.core.exception.OCommandExecutionException: Error on execution of command: sql.select *, out('Has')[@class=Tag][deleted_at=null] AS tags FROM Person
com.orientechnologies.orient.core.exception.OCommandExecutionException: Error on execution of command: sql.select *, out('Has')[@class=Tag][deleted_at=null] AS tags FROM Person
	at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.executeCommand(OAbstractPaginatedStorage.java:1190)
	at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.command(OAbstractPaginatedStorage.java:1173)
	at com.orientechnologies.orient.core.sql.query.OSQLQuery.run(OSQLQuery.java:71)
	at com.orientechnologies.orient.core.sql.query.OSQLSynchQuery.run(OSQLSynchQuery.java:85)
	at com.orientechnologies.orient.core.query.OQueryAbstract.execute(OQueryAbstract.java:33)
	at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.command(ONetworkProtocolBinary.java:1178)
	at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.executeRequest(ONetworkProtocolBinary.java:385)
	at com.orientechnologies.orient.server.network.protocol.binary.OBinaryNetworkProtocolAbstract.execute(OBinaryNetworkProtocolAbstract.java:216)
	at com.orientechnologies.common.thread.OSoftThread.run(OSoftThread.java:65)
Caused by: java.lang.NullPointerException
	at com.orientechnologies.orient.core.record.impl.ODocumentHelper.getFieldValue(ODocumentHelper.java:340)
	at com.orientechnologies.orient.core.sql.functions.coll.OSQLMethodMultiValue.execute(OSQLMethodMultiValue.java:54)
	at com.orientechnologies.orient.core.sql.methods.OSQLMethodRuntime.execute(OSQLMethodRuntime.java:139)
	at com.orientechnologies.orient.core.sql.filter.OSQLFilterItemAbstract.transformValue(OSQLFilterItemAbstract.java:140)
	at com.orientechnologies.orient.core.sql.functions.OSQLFunctionRuntime.execute(OSQLFunctionRuntime.java:128)
	at com.orientechnologies.orient.core.sql.ORuntimeResult.applyRecord(ORuntimeResult.java:119)
	at com.orientechnologies.orient.core.sql.ORuntimeResult.getProjectionResult(ORuntimeResult.java:225)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.addResult(OCommandExecutorSQLSelect.java:580)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.handleResult(OCommandExecutorSQLSelect.java:551)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.executeSearchRecord(OCommandExecutorSQLSelect.java:507)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.fetchFromTarget(OCommandExecutorSQLSelect.java:1316)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.executeSearch(OCommandExecutorSQLSelect.java:433)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.execute(OCommandExecutorSQLSelect.java:391)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLDelegate.execute(OCommandExecutorSQLDelegate.java:64)
	at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.executeCommand(OAbstractPaginatedStorage.java:1184)
	... 8 more

```

**Comment by tobiemh** [comment]
I've narrowed it down a little further:

The following query is successful indicating that it is a problem with records which only have 1 matching vertex:

```sql
SELECT *, out('Has')[@class=Tag][deleted_at=null][removed_at=null] AS tags FROM Person WHERE ( out('Has')[@class=Tag].size() > 1 OR out('Has')[@class=Tag].size() < 1 )
```

It might be as a result of this...

***

If I run `SELECT out('Has') AS tags FROM Person` the _tags_ are returned as a set...

![screen shot 2015-02-17 at 12 25 40](https://cloud.githubusercontent.com/assets/1588290/6228138/1d67fc78-b6a0-11e4-9d54-dfe7d0a5f835.png)

If I run `SELECT out('Has')[@class=Tag] AS tags FROM Person` the _tags_ are returned as a set if there are multiple _tags_ and a single rid if there is only one matching _tag_ ...

![screen shot 2015-02-17 at 12 27 49](https://cloud.githubusercontent.com/assets/1588290/6228170/79e7872a-b6a0-11e4-9f8e-6d5635fdd953.png)

Because it is not always returning a set, this is causing the problem.

The query `SELECT out('Has')[@class=Tag].asSet() AS tags FROM Person` shows the error a little more...

![screen shot 2015-02-17 at 12 30 05](https://cloud.githubusercontent.com/assets/1588290/6228211/de8b98d8-b6a0-11e4-97e2-091eb9d9901e.png)

OrientDB version: 2.0.2

**Comment by luigidellaquila** [comment]
Hi @tobiemh 

This problem is already fixed in current latest stable (2.2, but also 2.1).

I'm closing this issue, please reopen if you still see any related problems

Thanks

Luigi
